### PR TITLE
Add ECS/Fargate Support

### LINKED ISSUE #86: Fargate support
Hi,

I'm trying to deploy Hazelcast cluster on AWS Fargate but I couldn't find any documentation about the process...
Does Hazelcast suport Fargate or it only supports ECS?

All the best!

### LINKED ISSUE #97: dynamic configuration of host header and region parameters
In certain circumstances (e.g., Hazelcast cluster member using discovery, probably NOT clients) the `hostheader` and `region` can be detected automatically.

### LINKED ISSUE #98: ZONE_AWARE on ECS / Fargate
The ZONE_AWARE feature is not supported at the moment.
On Fargate a subnet identifies implicitly an availability zone, consider using this information to implement this.

### PR DESCRIPTION
**Pull request by leszko**
Implement ECS Fargate support.

fix #86
fix #97
fix #98 

Additional changes:
- Change EC2:DescibeInstances API calls to be consistent with ECS (parameters set with Headers, bot attributes (query params))
- Bump up Hamcrest library to `2.2` (#167)
- Refactor
- Add missing unit tests

### CODE REVIEW

File: src/main/java/com/hazelcast/aws/AwsClientConfigurator.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        AwsMetadataApi metadataApi = new AwsMetadataApi(awsConfig);
+
+        String region = resolveRegion(awsConfig, metadataApi, environment);
+        validateRegion(region);
+
+        AwsCredentialsProvider credentialsProvider = new AwsCredentialsProvider(awsConfig, metadataApi, environment);
+        AwsEc2Api ec2Api = createEc2Api(awsConfig, region);
+
+        // EC2 Discovery
+        if (explicitlyEc2Configured(awsConfig) || (!explicitlyEcsConfigured(awsConfig) && !environment.isRunningOnEcs())) {
```
Comment (pivovarit) [inline src/main/java/com/hazelcast/aws/AwsClientConfigurator.java:59]: Great job on the autodetection!

File: src/main/java/com/hazelcast/aws/AwsClientConfigurator.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            || ec2HostHeader.startsWith("ecs")
+            || ec2HostHeader.equals("ec2")
+        ) {
+            ec2HostHeader = DEFAULT_EC2_HOST_HEADER;
+        }
+        return ec2HostHeader.replace("ec2.", "ec2." + region + ".");
+    }
+
+    static String resolveEcsEndpoint(AwsConfig awsConfig, String region) {
+        String ecsHostHeader = awsConfig.getHostHeader();
```
Comment (pivovarit) [inline src/main/java/com/hazelcast/aws/AwsClientConfigurator.java:108]: Do we even need `host-header` anymore? Maybe we could make it optional and enum-like with a choice of either `ecs` or `ec2`?

File: src/main/java/com/hazelcast/aws/AwsClientConfigurator.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        if (isNotEmpty(awsConfig.getRegion())) {
+            return awsConfig.getRegion();
+        }
+
+        if (environment.isRunningOnEcs()) {
+            return environment.getAwsRegionOnEcs();
+        }
+
+        String availabilityZone = metadataApi.availabilityZoneEc2();
+        return availabilityZone.substring(0, availabilityZone.length() - 1);
```
[the remark this answers is not in the document]
Comment (leszko) [reply src/main/java/com/hazelcast/aws/AwsClientConfigurator.java]: Done

File: src/main/java/com/hazelcast/aws/AwsClientConfigurator.java (hunk 2 above)
Comment (leszko) [reply src/main/java/com/hazelcast/aws/AwsClientConfigurator.java:111]: As discussed on the stand up, I think we should keep it and consider removing (or changing to enum) in the next major release.

File: src/main/java/com/hazelcast/aws/Environment.java [hunk 4]
```diff
@@ -15,14 +15,25 @@
...
+    }
+
+    boolean isRunningOnEcs() {
+        return IS_RUNNING_ON_ECS;
+    }
 
-    String getEnv(String name) {
-        return System.getenv(name);
+    private static boolean isRunningOnEcsEnvironment() {
+        String execEnv = System.getenv("AWS_EXECUTION_ENV");
```
[the remark this answers is not in the document]
Comment (pivovarit) [reply src/main/java/com/hazelcast/aws/Environment.java:36]: Yes, these are set by AWS and can be used to do such lookups

File: src/main/java/com/hazelcast/aws/AwsCredentialsProvider.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        this.ec2IamRole = resolveEc2IamRole();
+    }
+
+    private String resolveEc2IamRole() {
+        if (StringUtils.isNotEmpty(awsConfig.getAccessKey())) {
+            // no need to resolve IAM Role, since using hardcoded Access/Secret keys takes precedence
+            return null;
+        }
+
+        if (StringUtils.isNotEmpty(awsConfig.getIamRole()) && !"DEFAULT".equals(awsConfig.getIamRole())) {
```
[the remark this answers is not in the document]
Comment (leszko) [reply src/main/java/com/hazelcast/aws/AwsCredentialsProvider.java:43]: It is not mentioned, I left it for the backwards-compatibility. I think it's not worth mentioning/using. Let's just remove it with the next major release.

File: src/main/java/com/hazelcast/aws/AwsClientConfigurator.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+    private static AwsEcsApi createEcsApi(AwsConfig awsConfig, String region) {
+        String ecsEndpoint = resolveEcsEndpoint(awsConfig, region);
+        AwsRequestSigner ecsRequestSigner = new AwsRequestSigner(region, ECS_SERVICE_NAME);
+        return new AwsEcsApi(ecsEndpoint, awsConfig, ecsRequestSigner, Clock.systemUTC());
+    }
+
+    static String resolveEc2Endpoint(AwsConfig awsConfig, String region) {
+        String ec2HostHeader = awsConfig.getHostHeader();
+        if (StringUtils.isEmpty(ec2HostHeader)
+            || ec2HostHeader.startsWith("ecs")
```
[the remark this answers is not in the document]
Comment (leszko) [reply src/main/java/com/hazelcast/aws/AwsClientConfigurator.java:102]: We need EC2 Endpoint for both discoveries (EC2 and ECS). That may seem a little counterintuitive, but ECS strategy also uses EC2 API (for discovering public IPs from EC2:DescribeNetworkInterfaces). So here, if someone didn't specified EC2 custom endpoint (Host Header does not equal to `ec2` or does not start from `ecs`), then we use the default EC2 endpoint.

**Pull request merged**